File: functions/src/services/aiProvider.ts
/**
 * AIProvider interface — ported from the FastAPI backend's `app/services/ai_provider.py`
 * verbatim in spirit. `NullAIProvider` is what ships until a real key is bound; it never
 * fakes a generated result, matching the platform's "no fake functionality" rule.
 */
export interface ProductContentRequest {
  kind: "description" | "tags" | "title";
  productName: string;
  category?: string | null;
  keywords?: string | null;
}

export interface ProductContentResult {
  configured: boolean;
  provider: string;
  message: string;
  content?: string;
}

export interface AIProvider {
  readonly name: string;
  isConfigured(): boolean;
  generateProductContent(req: ProductContentRequest): Promise<ProductContentResult>;
}

export class NullAIProvider implements AIProvider {
  readonly name = "none";
  isConfigured(): boolean {
    return false;
  }
  async generateProductContent(): Promise<ProductContentResult> {
    return {
      configured: false,
      provider: this.name,
      message:
        "AI Assistant isn't connected yet. Bind an AI_API_KEY secret and implement OpenAIProvider " +
        "in functions/src/services/aiProvider.ts to turn this on — until then this honestly reports " +
        "unconfigured instead of making something up.",
    };
  }
}

// Swap this for `new OpenAIProvider(apiKey)` (implement below) once AI_API_KEY is bound.
export function getAIProvider(apiKey: string | undefined): AIProvider {
  if (!apiKey) return new NullAIProvider();
  // TODO: implement OpenAIProvider here when a real key is available. Keeping this as an
  // explicit TODO rather than a stub that pretends to work is the whole point of this
  // interface — see ARCHITECTURE.md section 1, decision 5.
  return new NullAIProvider();
}

File: legacy-postgres-backend/app/services/ai_provider.py
"""AI Assistant provider interface.

This is architecture, not a mock: if no provider is configured, endpoints return a
clear "not configured" response instead of fabricating AI output. Wiring up a real
provider is a matter of implementing one more subclass and setting AI_PROVIDER/AI_API_KEY —
no endpoint or frontend code needs to change.
"""
import abc
from dataclasses import dataclass
from typing import Optional

from app.core.config import settings


@dataclass
class ProductContentRequest:
    product_name: Optional[str]
    category: Optional[str]
    keywords: Optional[str]
    tone: Optional[str] = "warm and friendly"


@dataclass
class ProductContentResult:
    title: str
    description: str
    tags: list
    social_caption: str


class AIProvider(abc.ABC):
    @abc.abstractmethod
    def is_configured(self) -> bool:
        raise NotImplementedError

    @abc.abstractmethod
    def generate_product_content(self, req: ProductContentRequest) -> ProductContentResult:
        raise NotImplementedError


class NullAIProvider(AIProvider):
    """Used when AI_PROVIDER=none. Every call raises clearly rather than returning
    fake content, per the platform's development rules."""

    def is_configured(self) -> bool:
        return False

    def generate_product_content(self, req: ProductContentRequest) -> ProductContentResult:
        raise RuntimeError(
            "AI Assistant is not configured yet. Set AI_PROVIDER and AI_API_KEY in your "
            "environment to enable AI-generated product content."
        )


class OpenAIProvider(AIProvider):
    """Real implementation placeholder — wire up the `openai` SDK here once a key is
    provided. Left unimplemented (not faked) until real credentials exist."""

    def __init__(self, api_key: str):
        self.api_key = api_key

    def is_configured(self) -> bool:
        return bool(self.api_key)

    def generate_product_content(self, req: ProductContentRequest) -> ProductContentResult:
        raise NotImplementedError(
            "OpenAIProvider is scaffolded but not yet implemented. Install the openai "
            "package and implement this method to enable live generation."
        )


def get_ai_provider() -> AIProvider:
    if settings.AI_PROVIDER == "openai" and settings.AI_API_KEY:
        return OpenAIProvider(settings.AI_API_KEY)
    return NullAIProvider()
